SourceCode/include/tracer/Ray.h

```diff
@@ -0,0 +1,18 @@
+#pragma once
+
+#include <optional>
+
+#include "Utils.h"
+#include "Vector.h"
+#include "tracer/Triangle.h"
+
+struct Ray {
+  constexpr static const float FLOAT_EPSILON = 1e-8;
+  Vector origin;
+  Vector direction;
+
+  Ray();
+  Ray(const Vector &origin, const Vector &direction);
+  std::optional<Vector> intersectWithTriangle(const Triangle &triangle) const;
```

Comment:
Usually it's better for this function to live in the Primitive that is intersected.

Reply:
Will the reason why be covered in future lectures of the course? If not, could you briefly explain?